GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,498 @@
+    end do ! i_point
+
+    deallocate (x_mu, psi, mat_A, weights_work, vec_S, mat_U, mat_VT, work, iwork, mat_VT_s, vec_UT_psi)
+
+  end subroutine get_transformation_weights
+
+  !> \brief Calculate the auxiliary matrix for cosine/sin transformation for tau to omega or viceversa
+  !! @param[in] num_points: Number of mesh points.
+  !! @param[in] tau_points: imaginary time grid points
+  !! @param[in] omega_points: imaginary frequency grid points
+  !! @param[in] num_x_nodes: Number of node in the interval [e_min,e_max]
+  !! @param[in] x_mu : Transition energy (nodes in the interval [e_min,e_max])
+  !! @param[inout] psi: corresponding auxiliary function (see transformation type definition)
+  !! @param[inout] mat_A: auxiliary matrix (see transformation type definition)  
+  !! @param[in] i_point: pointer for the current grid point
+  !! @param[inout] current_point:  current grid point ether omega(i_point) or tau_(i_point)
+  !! @param[in] transformation type :
+  !!        (1) the cosine transform cos(it) -> cos(iw): psi(omega,x), mat_A = cos(omega*tau)*psi(tau,x) 
+  !!        (2) the cosine transform cos(iw) -> cos(it): psi(tau,x)  , mat_A = cos(omega*tau)*psi(omega,x)
+  !!        (3) the sine transform   sin(it) -> sin(iw): psi(omega,x), mat_A = sin(omega*tau)*psi(tau,x)
+  subroutine calculate_psi_and_mat_A(num_points, tau_points, omega_points, num_x_nodes, x_mu, psi, &
+       mat_A, i_point, current_point, transformation_type)
+
+    integer, intent(in)                                :: num_points, num_x_nodes, i_point
+    real(kind=dp), allocatable, dimension(:), &
+         intent(in)                                    :: tau_points, omega_points, x_mu
+    real(kind=dp), allocatable, dimension(:), &          
+         intent(inout)                                 :: psi         
+    real(kind=dp), allocatable, dimension(:, :), &
+         intent(inout)                                 :: mat_A
+    real(kind=dp), intent(inout)                       :: current_point
+    integer, intent(in)                                :: transformation_type
+
+    ! Internal variables
+    integer                                            :: i_node, j_point
+    real(kind=dp)                                      :: tau, omega
+
+    ! Begin work
+
+    ! the cosine transform cos(it) -> cos(iw)
+    if (transformation_type.eq.1) then
+       omega = omega_points(i_point)
+       current_point = omega
+
+       ! psi(omega_k,x) = 2x/(x^2+omega_k^2)
+       do i_node = 1, num_x_nodes
+          psi(i_node) = 2.0_dp*x_mu(i_node)/((x_mu(i_node))**2 + omega**2)
+       end do
+
+       ! mat_A = cos(omega_k * tau) psi(tau,x)
+       do j_point = 1, num_points
+          tau = tau_points(j_point)
+          do i_node = 1, num_x_nodes
+             mat_A(i_node, j_point) = cos(omega*tau)*exp(-x_mu(i_node)*tau)
+          end do
+       end do
+
+    ! the cosine transform cos(iw) -> cos(it)  
+    else if (transformation_type.eq.2) then
```

Comment:
I recommend replacing numbers with named constants. Magic numbers are never good. 
For example:

```
module transform_types
   implicit none
   private
   !> cos(iw) -> cos(it) 
   integer, parameter, public :: cosine_transform = 0002
  ! Add others here
end module 

use transform_types, only: cosine_transform
...
else if (transformation_type.eq == cosine_transform) then
 ....
```

This also removes the need for the comment, the code becomes self-documenting.
It's also nice to a) not use .eq. operators apart from when dealing with logical comparisons and b) leaving a space between operators - reads nicer.

Reply:
Maybe @fdelesma could implement this change? I also think is a good idea

Reply:
I will take care :)

Reply:
Solved here e131f61949f8bf9371fe706952ffc242d4d0b83c